### PR TITLE
Bug 1812461 - Update CFR Popups to be sticky

### PR DESCRIPTION
**Pull request by DreVla**
CFRPopup default properties have been changed so a popup does not dismiss anymore on a click outside or a back press. Additionally, popups will reappear after a short delay if the screen is rotated and the popup was not dismissed.

https://user-images.githubusercontent.com/32488956/227920960-53eeb299-2aee-4588-8ed2-bf3bcc0fd332.mov

### Pull Request checklist

- [X] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [X] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1812461

### CODE REVIEW

**Comment by DreVla** [comment]
Previous commit migrated the `recommendPrivateBrowsing` PopupWindow into a CFRPopup in order to adhere to the new CFR Popup rules and make it future proof. Here are two picture of before and after and a video of the new behavior:

| Before PopupWindow | New CFRPopup |
| --- | --- |
| Screenshot 2023-03-27 at 12 36 43 | Screenshot 2023-03-27 at 12 36 11 |


https://user-images.githubusercontent.com/32488956/227928230-2925b673-6947-4f40-b1f1-5b2c7d39092b.mov

**Comment by boek** [comment]
@DreVla Is this ready to review?

**Comment by DreVla** [comment]
> @DreVla Is this ready to review?

Yes, now it is ready!

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/ContextualHintsTest.kt [hunk 1]
```diff
@@ -66,10 +64,10 @@ class ContextualHintsTest {
         val genericPage = getGenericAsset(mockWebServer, 1)
 
         navigationToolbar {
-        }.enterURLAndEnterToBrowser(genericPage.url) {
-            verifyCookiesProtectionHintIsDisplayed(true)
-            clickPageObject(itemWithDescription(getStringResource(R.string.mozac_cfr_dismiss_button_content_description)))
-            verifyCookiesProtectionHintIsDisplayed(false)
+        }.enterURLAndCheckTCPCFR(genericPage.url, composeTestRule) {
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/ContextualHintsTest.kt:67]: Same as above.

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/AddToHomeScreenRobot.kt [hunk 2]
```diff
@@ -5,35 +5,33 @@
...
-import org.mozilla.fenix.helpers.click
 import java.util.regex.Pattern
 
 /**
  * Implementation of Robot Pattern for the Add to homescreen feature.
  */
 class AddToHomeScreenRobot {
 
-    fun verifyAddPrivateBrowsingShortcutButton() = assertAddPrivateBrowsingShortcutButton()
+    fun verifyAddPrivateBrowsingShortcutButton(composeTestRule: ComposeTestRule) = assertAddPrivateBrowsingShortcutButton(composeTestRule)
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/AddToHomeScreenRobot.kt:30]: Let's do the check directly here and delete `assertAddPrivateBrowsingShortcutButton`

`fun verifyAddPrivateBrowsingShortcutButton(composeTestRule: ComposeTestRule) = composeTestRule.onNodeWithTag("private.add").assertIsDisplayed()`

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 3]
```diff
@@ -590,16 +591,22 @@ class BrowserRobot {
         }
     }
 
-    fun verifyCookiesProtectionHintIsDisplayed(isDisplayed: Boolean) {
-        assertItemContainingTextExists(
-            totalCookieProtectionHintMessage,
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:595]: The `totalCookieProtectionHintMessage` variable can be deleted as it's not used anymore.

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 4]
```diff
@@ -590,16 +591,22 @@ class BrowserRobot {
...
-            totalCookieProtectionHintLearnMoreLink,
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:596]: Same as above for `totalCookieProtectionHintLearnMoreLink`

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 5]
```diff
@@ -590,16 +591,22 @@ class BrowserRobot {
...
-            exists = isDisplayed,
-        )
-        assertItemWithDescriptionExists(
-            totalCookieProtectionHintCloseButton,
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:600]: Same as above for totalCookieProtectionHintCloseButton

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 6]
```diff
@@ -590,16 +591,22 @@ class BrowserRobot {
...
-            exists = isDisplayed,
-        )
+    fun verifyCookiesProtectionHintIsDisplayed(composeTestRule: HomeActivityComposeTestRule, isDisplayed: Boolean) {
+        if (isDisplayed) {
+            composeTestRule.onNodeWithTag("tcp_cfr.message").assertIsDisplayed()
+            composeTestRule.onNodeWithTag("tcp_cfr.action").assertIsDisplayed()
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:597]: The "dismiss" button check seems to be missing.
`composeTestRule.onNodeWithTag("cfr.dismiss").assertIsDisplayed()`

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 7]
```diff
@@ -590,16 +591,22 @@ class BrowserRobot {
...
+        } else {
+            composeTestRule.onNodeWithTag("tcp_cfr.message").assertDoesNotExist()
+            composeTestRule.onNodeWithTag("tcp_cfr.action").assertDoesNotExist()
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:600]: The check that verifies that the "dismiss" button is not displayed, seems to be missing.
`composeTestRule.onNodeWithTag("cfr.dismiss").assertDoesNotExist()`

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 8]
```diff
@@ -590,16 +591,22 @@ class BrowserRobot {
...
+        }
+    }
+
+    fun clickTCPCFRLearnMore(composeTestRule: HomeActivityComposeTestRule) {
+        composeTestRule.onNodeWithTag("tcp_cfr.action").performClick()
+    }
+
+    fun dismissCFRPopup(composeTestRule: HomeActivityComposeTestRule) {
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:608]: Is this specific to all CFRs ?
If it's only for the TCP CFR, I would rename it.

File: fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 9]
```diff
@@ -993,11 +1056,6 @@ class HomeFragment : Fragment() {
         private const val SCROLL_TO_COLLECTION = "scrollToCollection"
         private const val ANIM_SCROLL_DELAY = 100L
 
-        private const val CFR_WIDTH_DIVIDER = 1.7
-        private const val CFR_Y_OFFSET = -20
```
Comment (Alexandru2909) [inline fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:997]: We can account for this offset using `CFRPopupProperties.popupVerticalOffset`.

File: fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 10]
```diff
@@ -826,49 +838,100 @@ class HomeFragment : Fragment() {
...
+                                .semantics {
+                                    testTagsAsResourceId = true
+                                    testTag = "private.message"
+                                },
+                        )
+                    }
+                },
+                action = {
+                    FirefoxTheme {
+                        Button(
```
Comment (Alexandru2909) [inline fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:882]: Could this also be a `TextButton`?

File: fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 11]
```diff
@@ -826,49 +838,100 @@ class HomeFragment : Fragment() {
...
+                            },
+                            colors = ButtonDefaults.buttonColors(backgroundColor = PhotonColors.LightGrey30),
+                            modifier = Modifier
+                                .padding(top = 16.dp)
+                                .heightIn(36.dp)
+                                .fillMaxWidth()
+                                .semantics {
+                                    testTagsAsResourceId = true
+                                    testTag = "private.add"
+                                },
```
Comment (Alexandru2909) [inline fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:898]: For better consistency with the popup, we could also specify the shape of this button as `RoundedCornerShape(8.dp)`, as specified in `rounded_gray_corners.xml`

**Review by Alexandru2909 — COMMENTED** [review]
This looks very good! I have a couple of questions about the CFR behavior and assuming the automated test feedback is addressed, this should be good to go.

File: android-components/components/compose/cfr/src/main/java/mozilla/components/compose/cfr/CFRPopupContent.kt
Comment (Alexandru2909) [inline android-components/components/compose/cfr/src/main/java/mozilla/components/compose/cfr/CFRPopupContent.kt:1]: I can't add a comment there, but on line 112 we have `end = 24.dp, // 8.dp extra padding to the "X" icon`. Should we still account for this when the dismiss button is now displayed?

Comment (DreVla) [reply android-components/components/compose/cfr/src/main/java/mozilla/components/compose/cfr/CFRPopupContent.kt:1]: Tbh not sure if there is a possibility of having a CFR without a "X" dismiss button, but just to be sure I added an extra check there. 👍

**Review by AndiAJ — APPROVED** [review]
👍🏽

**Review by Alexandru2909 — APPROVED** [review]
LGTM! Nice work

**Pull request merged**